### PR TITLE
Added periodic xmpp ping functionality.

### PR DESCRIPTION
**Pull request by bolshoy**
NOTE: xmpp package patched (not yet forked) to expose JID for ping.
Added gracefulClose func that waits until all messages are processed and then closes the xmpp client.

### CODE REVIEW

File: gcm.go [hunk 1]
```diff
@@ -263,13 +264,61 @@ func newXmppGcmClient(senderID string, apiKey string) (*xmppGcmClient, error) {
...
+// ping sends a ping message and blocks until pong is received.
+//
+// Returns error if timeout time passes before pong.
+func (c *xmppGcmClient) ping(timeout time.Duration) error {
+	if err := c.XmppClient.PingC2S(c.XmppClient.JID(), xmppHost); err != nil {
+		return err
+	}
+	select {
+	case <-c.pongs:
+		//if pongID == pingID {
```
Comment (oryband) [inline gcm.go]: ?

File: gcm.go [hunk 2]
```diff
@@ -335,10 +387,18 @@ func (c *xmppGcmClient) listen(h MessageHandler, stop <-chan bool) error {
 			// This is likely fatal, so return.
 			return fmt.Errorf("error on Recv>%v", err)
 		}
-		v, ok := stanza.(xmpp.Chat)
-		if !ok {
+		switch stanza.(type) {
+		case xmpp.Chat:
```
Comment (oryband) [inline gcm.go]: having a case with no statements and no `default:` case means that this is a no-op. you can remove this line

Comment (oryband) [inline gcm.go]: unless you wanted to do a `fallthrough` in which case this is not falling thorough

File: gcm.go (hunk 1 above)
Comment (bolshoy) [inline gcm.go]: just a reminded that ping tracking can be implemented here.

**Pull request merged**